src/os/PluggableDBStore.cc

```diff
@@ -0,0 +1,481 @@
+
+    if (!dl) {
+       derr <<" Error opening the library" << err << dendl;
+       return -1;
+    }
+	library = dl;
+    
+    int n = numelems(table);
+    for (i = 0; i < n; i++) {
+        const char *name = table[i].name;
+        void *func = dlsym(dl, name);
+        if (func) {
+            *(void **)table[i].func = func;
+	    } else {
+	       derr << "Error loading symbol: " << name << dendl;	
+    	   err = dlerror();
+           derr <<" Error loading symbol: " << err << dendl;
+	       return -1;
+	    }
+    }
+    return 0;
+}
+
+int PluggableDBStore::_test_init()
+{
+  dout(10) << __func__ << dendl;
+  PluggableDBStore::create_if_missing = true;
+  return 0;
+}
+
+int PluggableDBStore::getosdid()
+{
+  dout(30) << __func__ << dendl;
+  char *end;
+  const char *id = g_conf->name.get_id().c_str();
+  int whoami = strtol(id, &end, 10);
+  return whoami;
+}
+
+int PluggableDBStore::init(string opt_str)
+{
+  dout(10) << __func__ << opt_str << dendl;
+  int ret = load_symbols(g_conf->keyvaluestore_backend_library.c_str());
+  assert(ret ==0);
+  const char* path = g_conf->osd_data.data();
+  map<string, string> options;
+  if (PluggableDBStore::create_if_missing) {
+    options["create_if_missing"] = "true";
+  } else {
+    options["create_if_missing"] = "false";
+  }
+  int status = DBInit(path, getosdid(), &options);
+  assert(status);
+  init_done = true;
+  return 0;
+}
+
+int PluggableDBStore::do_open(ostream &out, bool create_if_missing)
+{
+  dout(10) << __func__ <<dendl;
+  if (!init_done) {
+    init(string());
+  }
+  PerfCountersBuilder plb(g_ceph_context, "db", l_PluggableDB_first, l_PluggableDB_last);
+  plb.add_u64_counter(l_PluggableDB_gets, "db_get");
+  plb.add_u64_counter(l_PluggableDB_txns, "db_transaction");
+  logger = plb.create_perf_counters();
+  cct->get_perfcounters_collection()->add(logger);
+  return 0;
+}
+
+PluggableDBStore::PluggableDBStore(CephContext *c) :
+  cct(c),
+  logger(NULL),init_done(false)
+{
+}
+
+PluggableDBStore::~PluggableDBStore()
+{
+  close();
+  dlclose(library);
+  delete logger;
+}
+
+void PluggableDBStore::close()
+{
+  //shutdown the DB backend
+  DBClose();
+  if (logger)
+    cct->get_perfcounters_collection()->remove(logger);
+}
+
+int PluggableDBStore::submit_transaction(KeyValueDB::Transaction t)
+{
+  dout(20) << __func__ <<dendl;
+  PluggableDBTransactionImpl * _t =
+    static_cast<PluggableDBTransactionImpl *>(t.get());
+
+  int ret = DBSubmitTransaction(_t->ops);
+  logger->inc(l_PluggableDB_txns);
+  return ret;
+}
+
+int PluggableDBStore::submit_transaction_sync(KeyValueDB::Transaction t)
+{
+  return submit_transaction(t);
+}
+
+void PluggableDBStore::PluggableDBTransactionImpl::set(
+  const string &prefix,
+  const string &k,
+  const bufferlist &to_set_bl)
+{
+  string key = combine_strings(prefix, k);
+  DBOp op;
+  op.type = DB_OP_WRITE;
+  string data;
+  to_set_bl.copy(0,to_set_bl.length(), data);
+  op.data = data;
+  ops[key] = op;
```

Comment:
we can have multiple buffers in the bufferlist we get here, if we use c_str() it will call rebuild and memcpy is involved there also. So to have a contiguous allocation, copying to a string. will change this to a pointer value, instead of string to avoid a copy.